### PR TITLE
Service service2

### PR DESCRIPTION
**Pull request by guilhermevanzin**
Create ServiceService and SchedulingService

### CODE REVIEW

File: Sellers.Services/SchedulingService.cs [hunk 1]
```diff
@@ -4,13 +4,148 @@
...
+        
+        public async Task<SchedulingViewModel> CreateAsync(SchedulingInputModel inputModel)
+        {
+            if (inputModel == null)
+                throw new ArgumentNullException(nameof(inputModel));
+
+            var scheduligns = await _schedulingRepository.GetAllAsync();
+
+            if(scheduligns.Any(s => s.ScheduleDate.ToString().Equals(inputModel.ScheduleDate)))
+               throw new ArgumentException($"{inputModel.ScheduleDate}: Date not disponible");
```
Comment (JulioDeBastiani) [inline Sellers.Services/SchedulingService.cs]: Fix to account for service duration

File: Sellers.Domain/Models/Scheduling.cs [hunk 2]
```diff
@@ -31,7 +34,13 @@ public Scheduling(uint id, Service service, Client client, Seller seller, string
...
+            SetScheduleDate(scheduleDate);
+
+            SetServiceDuration(service.Duration);
+
+            SetPercCommission(seller.ServiceCommission);
+
+            SetPrice(service.Price);    
         }
 
         public void SetScheduleDate(String scheduleDate)
```
Comment (JulioDeBastiani) [inline Sellers.Domain/Models/Scheduling.cs]: should take a DateTime and not a stirng

File: Sellers.Services/SchedulingService.cs [hunk 3]
```diff
@@ -4,13 +4,148 @@
...
+            {
+                var scheduligns = await _schedulingRepository.GetAllAsync();
+
+                if(scheduligns.Any(s => s.ScheduleDate.ToString().Equals(updateModel.ScheduleDate)))
+                    throw new ArgumentException($"{updateModel.ScheduleDate}: Date not disponible");
+                
+                scheduling.SetScheduleDate(updateModel.ScheduleDate.ToString());
+            }
+
+            if(updateModel.SellerId != 0)
```
Comment (JulioDeBastiani) [inline Sellers.Services/SchedulingService.cs]: Use null instead of 0

File: Sellers.Services/SchedulingService.cs [hunk 4]
```diff
@@ -4,13 +4,148 @@
...
+            {
+                var seller = await _sellerRepository.GetAsync(scheduling.SellerId);
+
+                if(seller == null)
+                    throw new ArgumentNullException(nameof(seller));
+                
+                scheduling.SetSeller(seller);
+            }
+            
+            if(updateModel.ClientId != 0)
```
Comment (JulioDeBastiani) [inline Sellers.Services/SchedulingService.cs]: use null instead of 0

File: Sellers.Services/SchedulingService.cs [hunk 5]
```diff
@@ -4,13 +4,148 @@
...
+            {
+                var client = await _clientRepository.GetAsync(scheduling.ClientId);
+
+                if(client == null)
+                    throw new ArgumentNullException(nameof(client));
+                
+                scheduling.SetClient(client);
+            }
+
+            if(updateModel.ServiceId != 0)
```
Comment (JulioDeBastiani) [inline Sellers.Services/SchedulingService.cs]: use null instead of 0

File: Sellers.Domain/Models/Scheduling.cs (hunk 2 above)
Comment (guilhermevanzin) [reply Sellers.Domain/Models/Scheduling.cs]: Fix in 88c3e5e7719b5411ff3411c2f388ba705c42f5fd

File: Sellers.Services/SchedulingService.cs (hunk 1 above)
Comment (guilhermevanzin) [reply Sellers.Services/SchedulingService.cs]: Fix in 88c3e5e7719b5411ff3411c2f388ba705c42f5fd

File: Sellers.Domain/Models/Scheduling.cs [hunk 6]
```diff
@@ -69,14 +78,28 @@ public void SetSeller(Seller seller)
             Seller = seller;
         }
 
-        public void SetCommission(double commission)
+        public void SetServiceDuration(uint duration)
```
[the remark this answers is not in the document]
Comment (guilhermevanzin) [reply Sellers.Domain/Models/Scheduling.cs]: Fix in 88c3e5e7719b5411ff3411c2f388ba705c42f5fd

File: Sellers.Domain/Models/Scheduling.cs [hunk 7]
```diff
@@ -69,14 +78,28 @@ public void SetSeller(Seller seller)
...
+        public void SetPrice(double price)
         {
-            throw new NotImplementedException();
+            if(price < 0)
+                throw new ArgumentOutOfRangeException(nameof(price));
+            
+            Price = price;
+        }
+        
+        public void SetPercCommission(double percCommission)
```
[the remark this answers is not in the document]
Comment (guilhermevanzin) [reply Sellers.Domain/Models/Scheduling.cs]: Fix in d477d51f5c573f7435472ab2d4243d6eb66baf0f

File: Sellers.Services/SchedulingService.cs (hunk 3 above)
Comment (guilhermevanzin) [reply Sellers.Services/SchedulingService.cs]: Fix in 44b13df1dcfcc15b81d6480122175a6f0e064a19

File: Sellers.Services/SchedulingService.cs (hunk 4 above)
Comment (guilhermevanzin) [reply Sellers.Services/SchedulingService.cs]: Fix in 44b13df1dcfcc15b81d6480122175a6f0e064a19

File: Sellers.Services/SchedulingService.cs (hunk 5 above)
Comment (guilhermevanzin) [reply Sellers.Services/SchedulingService.cs]: Fix in 44b13df1dcfcc15b81d6480122175a6f0e064a19

File: Sellers.Domain/Models/Service.cs [hunk 8]
```diff
@@ -57,12 +57,15 @@ public void SetPrice(double price)
             Price = price;
         }
 
-        public void SetDuration(uint duration)
+        public void SetDuration(int? duration)
```
Comment (JulioDeBastiani) [inline Sellers.Domain/Models/Service.cs]: this param has no reason to be nullable, also either change nomenclature os change type to TimeSpan

File: Sellers.Domain/Models/Service.cs [hunk 9]
```diff
@@ -57,12 +57,15 @@ public void SetPrice(double price)
...
-            if(duration <= 0)
-                throw new ArgumentOutOfRangeException(nameof(duration));
+            if(duration == null)
+                throw new ArgumentNullException(nameof(duration));
             
-            Duration = duration;
+            if(duration < 1)
+                 throw new ArgumentOutOfRangeException(nameof(duration));
+            
+            Duration = new TimeSpan(duration.Value,0,0);
```
Comment (JulioDeBastiani) [inline Sellers.Domain/Models/Service.cs]: Use TimeSpan.FromX instead of the constructor

File: Sellers.Services/InputModels/ServiceInputModel.cs [hunk 10]
```diff
@@ -6,6 +7,6 @@ public class ServiceInputModel
         public string BriefDescr { get; set; }
         public string DetailedDesc { get; set; }
         public double Price { get; set; }
-        public uint Duration { get; set; } 
+        public int? Duration { get; set; }
```
Comment (JulioDeBastiani) [inline Sellers.Services/InputModels/ServiceInputModel.cs]: Either change to `TimeSpan Duration` or `int DurationHours`

File: Sellers.Services/SchedulingService.cs [hunk 11]
```diff
@@ -56,6 +51,21 @@ public async Task<SchedulingViewModel> CreateAsync(SchedulingInputModel inputMod
...
+            var endHourInputModel = inputModel.ScheduleDate.AddHours(service.Duration.Hours).Hour;
+                
+            foreach (var tempScheduling in schedulings)
+            {
+                var startHour  = tempScheduling.ScheduleDate.Hour;
+                var endHour = tempScheduling.ScheduleDate.AddHours(tempScheduling.ServiceDuration.Hours).Hour;
+                    
+                if((startHourInputModel >= startHour && startHourInputModel <= endHour) || (endHourInputModel >= startHour && endHourInputModel <= endHour))
+                    throw new ArgumentException($"Scheduling {tempScheduling.Id} already exists on this hour range");
+            }
```
Comment (JulioDeBastiani) [inline Sellers.Services/SchedulingService.cs]: This should be a method

File: Sellers.Services/SchedulingService.cs [hunk 12]
```diff
@@ -89,49 +99,59 @@ public virtual async Task<SchedulingViewModel> PutAsync(uint id, SchedulingUpdat
...
+                var endHourUpdateModel = updateModel.ScheduleDate.AddHours(updateModel.Duration.Hours).Hour;
+                    
+                foreach (var tempScheduling in schedulings)
+                {
+                    var startHour  = tempScheduling.ScheduleDate.Hour;
+                    var endHour = tempScheduling.ScheduleDate.AddHours(tempScheduling.ServiceDuration.Hours).Hour;
+                     
+                    if((startHourUpdateModel >= startHour && startHourUpdateModel <= endHour) || (endHourUpdateModel >= startHour && endHourUpdateModel <= endHour))
+                        throw new ArgumentException($"Scheduling {tempScheduling.Id} already exists on this hour range");
+                }
```
Comment (JulioDeBastiani) [inline Sellers.Services/SchedulingService.cs]: This should be a method

File: Sellers.Services/ViewModels/ServiceViewModel.cs [hunk 13]
```diff
@@ -8,7 +9,7 @@ public class ServiceViewModel
         public string BriefDescr { get; set; }
         public string DetailedDescr { get; set; }
         public double Price { get; set; }
-        public uint Duration { get; set; }
+        public int Duration { get; set; }
```
Comment (JulioDeBastiani) [inline Sellers.Services/ViewModels/ServiceViewModel.cs]: This has no reason not to be a TimeSpan

File: Sellers.Domain/Models/Service.cs (hunk 8 above)
Comment (guilhermevanzin) [reply Sellers.Domain/Models/Service.cs]: Fix in ea5a5455ac5ebc1e393d44297f583a522ce99baa

File: Sellers.Domain/Models/Service.cs (hunk 9 above)
Comment (guilhermevanzin) [reply Sellers.Domain/Models/Service.cs]: Fix in ea5a5455ac5ebc1e393d44297f583a522ce99baa

File: Sellers.Services/InputModels/ServiceInputModel.cs (hunk 10 above)
Comment (guilhermevanzin) [reply Sellers.Services/InputModels/ServiceInputModel.cs]: Fix in ea5a5455ac5ebc1e393d44297f583a522ce99baa

File: Sellers.Services/SchedulingService.cs (hunk 11 above)
Comment (guilhermevanzin) [reply Sellers.Services/SchedulingService.cs]: Fix in ea5a5455ac5ebc1e393d44297f583a522ce99baa

File: Sellers.Services/SchedulingService.cs (hunk 12 above)
Comment (guilhermevanzin) [reply Sellers.Services/SchedulingService.cs]: Fix in ea5a5455ac5ebc1e393d44297f583a522ce99baa

File: Sellers.Services/ViewModels/ServiceViewModel.cs (hunk 13 above)
Comment (guilhermevanzin) [reply Sellers.Services/ViewModels/ServiceViewModel.cs]: Fix in ea5a5455ac5ebc1e393d44297f583a522ce99baa

File: Sellers.Services/SchedulingService.cs [hunk 14]
```diff
@@ -167,5 +149,23 @@ public virtual async Task DeleteAsync(uint id)
...
+             var startHourInput = scheduleHour;
+             var endHourInput = scheduleHour.Add(serviceDuration);
+                
+            foreach (var tempScheduling in schedulings)
+            {
+                var startHour  = tempScheduling.ScheduleDate.TimeOfDay;
+                var endHour = startHour.Add(tempScheduling.ServiceDuration);
+                    
+                if((startHourInput >= startHour && startHourInput <= endHour) || (endHourInput >= startHour && endHourInput <= endHour))
+                    throw new ArgumentException($"Scheduling {tempScheduling.Id} already exists on this hour range");
```
Comment (JulioDeBastiani) [inline Sellers.Services/SchedulingService.cs]: Return a list of schedulings instead of throwing an exception

Comment (guilhermevanzin) [reply Sellers.Services/SchedulingService.cs]: Fix in 7409de6c487ecadf120f4c764290ade7b55605f7

File: Sellers.Services/SchedulingService.cs [hunk 15]
```diff
@@ -150,22 +167,14 @@ public virtual async Task DeleteAsync(uint id)
             await _schedulingRepository.RemoveAsync(scheduling);
         }
 
-        private async Task ValidateScheduleDate(DateTime scheduleDate,TimeSpan scheduleHour,TimeSpan serviceDuration)
+        private async Task<IEnumerable<Scheduling>> ValidateScheduleDate(DateTime Date, TimeSpan serviceDuration)
```
Comment (JulioDeBastiani) [inline Sellers.Services/SchedulingService.cs]: Update the method's name

Comment (guilhermevanzin) [reply Sellers.Services/SchedulingService.cs]: Fix in 7a71bd84c20959d04cfb6e125023634d18bcc07d

**Pull request merged**